include/mitsuba/render/sampler.h

```diff
@@ -41,20 +88,54 @@ class MTS_EXPORT_RENDER Sampler : public Object {
     virtual Point2f next_2d(Mask active = true);
 
     /// Return the number of samples per pixel
-    size_t sample_count() const { return m_sample_count; }
+    uint32_t sample_count() const { return m_sample_count; }
 
     /// Return the size of the wavefront (or 0, if not seeded)
-    virtual size_t wavefront_size() const = 0;
+    uint32_t wavefront_size() const;
+
+    /// Set the number of samples per pass in wavefront modes (default is 1)
+    void set_samples_per_wavefront(uint32_t samples_per_wavefront);
 
     MTS_DECLARE_CLASS()
 protected:
     Sampler(const Properties &props);
     virtual ~Sampler();
 
+    /// Generates a array of seeds where the seed values are unique per sequence
+    UInt32 compute_per_sequence_seed(uint64_t seed_offset) const;
+    /// Return the index of the current sample
+    UInt32 current_sample_index() const;
+
+protected:
+    uint64_t m_base_seed;
+    uint32_t m_sample_count;
+    uint32_t m_samples_per_wavefront;
+    /// Array of offsets to compute current sample index in wavefront modes
+    UInt32 m_wavefront_sample_offsets;
```

Comment:
Is there any fundamental reason that this needs to be stored as a variable? In practice, this computation should be merged into a kernel and then recomputed at every wavefront. Paying the cost of reading a variable just to avoid a division or modulo is not a good trade-off for the GPU.

Reply:
I will replace it with `uint32_t m_wavefront_size` then.